pkg/daemon/ceph/client/upgrade.go

```diff
@@ -18,46 +18,99 @@ package client
+	Mgr       map[string]int `json:"mgr,omitempty"`
+	Osd       map[string]int `json:"osd,omitempty"`
+	Rgw       map[string]int `json:"rgw,omitempty"`
+	Mds       map[string]int `json:"mds,omitempty"`
+	RbdMirror map[string]int `json:"rbd-mirror,omitempty"`
+	Overall   map[string]int `json:"overall,omitempty"`
 }
 
-func getCephMonVersionString(context *clusterd.Context) (string, error) {
-	output, err := context.Executor.ExecuteCommandWithOutput(false, "", "ceph", "version")
+var (
+	// we don't perform any checks on these daemons
+	// they don't have any "ok-to-stop" command implemented
+	daemonNoCheck = []string{"mgr", "rgw", "rbd-mirror", "nfs"}
+)
+
+func getCephMonVersionString(context *clusterd.Context, clusterName string) (string, error) {
+	args := []string{"version"}
+	command, args := FinalizeCephCommandArgs("ceph", args, context.ConfigDir, clusterName)
+
+	output, err := context.Executor.ExecuteCommandWithOutput(false, "", command, args...)
+	if err != nil {
+		return "", fmt.Errorf("failed to run ceph version. %+v", err)
+	}
+	logger.Debug(output)
+
+	return output, nil
+}
+
+func getCephVersionsString(context *clusterd.Context, clusterName string) (string, error) {
+	args := []string{"versions"}
+	command, args := FinalizeCephCommandArgs("ceph", args, context.ConfigDir, clusterName)
+
+	output, err := context.Executor.ExecuteCommandWithOutput(false, "", command, args...)
 	if err != nil {
-		return "", fmt.Errorf("failed to run ceph version: %+v", err)
+		return "", fmt.Errorf("failed to run ceph versions. %+v", err)
 	}
 	logger.Debug(output)
 
 	return output, nil
 }
 
-func getCephVersionsString(context *clusterd.Context) (string, error) {
-	output, err := context.Executor.ExecuteCommandWithOutput(false, "", "ceph", "versions")
+func getCephDaemonVersionString(context *clusterd.Context, deployment, clusterName string) (string, error) {
+	daemonName, err := findDaemonName(deployment)
 	if err != nil {
-		return "", fmt.Errorf("failed to run ceph versions: %+v", err)
+		return "", fmt.Errorf("%+v", err)
+	}
+	daemonID := findDaemonID(deployment)
+	daemon := daemonName + "." + daemonID
+
+	args := []string{"tell", daemon, "version"}
+	command, args := FinalizeCephCommandArgs("ceph", args, context.ConfigDir, clusterName)
+	output, err := context.Executor.ExecuteCommandWithOutput(false, "", command, args...)
+	if err != nil {
+		return "", fmt.Errorf("failed to run ceph tell. %+v", err)
 	}
 	logger.Debug(output)
 
 	return output, nil
 }
 
 // GetCephMonVersion reports the Ceph version of all the monitors, or at least a majority with quorum
-func GetCephMonVersion(context *clusterd.Context) (*cephver.CephVersion, error) {
-	output, err := getCephMonVersionString(context)
+func GetCephMonVersion(context *clusterd.Context, clusterName string) (*cephver.CephVersion, error) {
+	output, err := getCephMonVersionString(context, clusterName)
+	if err != nil {
+		return nil, fmt.Errorf("failed to run ceph version. %+v", err)
```

Comment:
This is repeating the same error text from `getCephMonVersionString`. This should be adding text for understanding the request in the context of `GetCephMonVersion`. I'm not sure how much extra context is needed here, so I would probably just return `err`.

Reply:
bump